src/types/entities/schelling.ts

```diff
@@ -37,72 +37,142 @@ export class SchellingGame {
+					round.render()
+				}
+			}
+			this.currentRoundNo = roundNo
+		}
+		this.lastBlock = block
+		const blocksPerRound = config.blocksPerRound // TODO use configured blocksPerRound
+		const offset = block.blockNo % blocksPerRound
+		let phase
+		let length
+		let elapsed
+		if (offset < blocksPerRound / 4) {
+			phase = 'commit'
+			length = blocksPerRound / 4
+			elapsed = offset + 1
+		} else if (offset <= blocksPerRound / 2) {
+			phase = 'reveal'
+			length = blocksPerRound / 4 + 1
+			elapsed = offset - blocksPerRound / 4 + 1
+		} else {
+			phase = 'claim'
+			length = blocksPerRound / 2 - 1
+			elapsed = offset - blocksPerRound / 2
+		}
+		const remaining = length - elapsed
+		const percent = Math.floor((elapsed * 100) / length)
+
+		let line = `${Round.roundString(
+			block.blockNo
+		)} ${percent}% of ${phase}, ${remaining} blocks left`
+		if (config.blocksPerRound - offset - 1 != remaining)
+			line = line + `, ${config.blocksPerRound - offset - 1} in round`
+		return line
+	}
 
+	public getOrCreateRound(roundNo: number, block: BlockDetails): Round {
 		// create the round if it doesn't exist
 		if (!this.rounds.has(roundNo)) {
 			this.rounds.set(roundNo, new Round(block))
-		} else {
-			// update the round timestamp
-			this.rounds.get(roundNo)!.lastBlock = block
+			this.players.forEachPair((overlay, player) => {
+				player.notPlaying()
+			})
+		}
+		const round = this.rounds.get(roundNo)
+		round!.lastBlock = block
+		return round!
+	}
+
+	// --- commit
+	public commit(overlay: string, owner: string, block: BlockDetails) {
+		const roundNo = SchellingGame.roundFromBlockNo(block.blockNo)
+
+		const round = this.getOrCreateRound(roundNo, block)
+
+		// create the player if it doesn't exist
+		if (!this.players.has(overlay)) {
+			this.players.set(overlay, new Player(overlay, owner, block, this.size))
 		}
 
 		// update player state
 		const player = this.players.get(overlay)!
 		player.commit(block)
 
 		// update the round state
-		const round = this.rounds.get(roundNo)!
 		round.commits++
 		round.players.push(overlay)
 
-		Logging.showError(
-			`${Round.roundString(block.blockNo)} Player ${leftId(overlay, 16)}`
-		)
+		if (SchellingGame._showLogs)
+			Logging.showError(
+				`${Round.roundString(block.blockNo)} Player ${leftId(overlay, 16)}`
+			)
 
 		round.render()
 	}
 
 	// --- reveal
 	public reveal(
 		overlay: string,
+		owner: string,
 		hash: string,
 		depth: number,
 		block: BlockDetails
 	) {
 		const roundNo = SchellingGame.roundFromBlockNo(block.blockNo)
 
-		// create the round if it doesn't exist
-		if (!this.rounds.has(roundNo)) {
-			this.rounds.set(roundNo, new Round(block))
-		} else {
-			// update the round timestamp
-			this.rounds.get(roundNo)!.lastBlock = block
+		const round = this.getOrCreateRound(roundNo, block)
+
+		// create the player if it doesn't exist
+		if (!this.players.has(overlay)) {
+			this.players.set(overlay, new Player(overlay, owner, block, this.size))
```

Comment:
This should actually be a violation of the invariants - if we maintain the invariant that we only start syncing from the start of a round, it should be impossible to call this function without calling `claim` prior, so there should be no need to have a `getOrCreateRound` - we should just always create it in `commit`

Reply:
My suggestion for this is enforcing the invariant at the start of the `reveal` and `claim` function that the player must be playing (and that the round must exist).

Reply:
I don't see any problems with just picking up new players in whatever state they first appear, but I can be persuaded to enforce an invariant if you can show a down-side to the late discovery of new players.